File: Answer/396.旋转函数.py
#
# @lc app=leetcode.cn id=396 lang=python3
#
# [396] 旋转函数
#

# @lc code=start
from typing import List


class Solution:
    def maxRotateFunction(self, nums: List[int]) -> int:
        n = len(nums)

        total, f = 0, 0
        for i, v in enumerate(nums):
            total += v
            f += i * v

        max_f = f
        for i in range(n - 1, 0, -1):
            f += total - n * nums[i]
            if f > max_f:
                max_f = f
        return max_f
# @lc code=end

File: Answer/396.旋转函数.cs
/*
 * @lc app=leetcode.cn id=396 lang=csharp
 *
 * [396] 旋转函数
 */

// @lc code=start
public class Solution {
    public int MaxRotateFunction(int[] nums) {
        int n = nums.Length;

        int sum = 0, f = 0;
        for (int i = 0; i < n; i++) {
            sum += nums[i];
            f += i * nums[i];
        }

        int maxF = f;
        for (int i = n - 1; i > 0; i--) {
            f += sum - n * nums[i];
            if (f > maxF) {
                maxF = f;
            }
        }
        return maxF;
    }
}
// @lc code=end
